main/src/ca/uwaterloo/flix/language/phase/unification/FastSetUnification.scala

```diff
@@ -0,0 +1,338 @@
+/*
+ * Copyright 2024 Jonathan Lindegaard Starup
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase.unification
+
+import scala.annotation.nowarn
+import scala.collection.immutable.SortedSet
+
+object FastSetUnification {
+
+  /**
+    * A common super-type for set formulas `f`, like `x1 ∩ x2 ∪ (e4 ∪ !c17)`.
+    *
+    * A set formula can contain elements, constants (fixed unknown sets), variables (unknown sets),
+    * complements, unions, and intersections.
+    *
+    * Variables and constants are collectively referred to as unknowns.
+    * A formula without unknowns is called ground.
+    *
+    * A set without unknowns can be evaluated to get a finite or a co-finite set of elements. The
+    * universe of elements is infinite so no finite set is equivalent to universe.
+    *
+    * Invariant: [[SetFormula.ElemSet]], [[SetFormula.Cst]], and [[SetFormula.Var]] must use
+    * disjoint integers (see [[unknowns]]).
+    */
+  sealed trait SetFormula {
+
+    import SetFormula.*
+
+    /** The set of [[Var]] in `this`. */
+    final def variables: SortedSet[Int] = this match {
```

Comment:
May I propose freeVars? (We use that terminology elsewhere).

Reply:
I think its a bit misleading when there is no of bound/free distinction here, variables are just variables

Reply:
OK. We keep it then. But in most places we use freeVars irrespective of whether there are binders.